File: 0056-merge-intervals/0056-merge-intervals.cpp
class Solution {
public:
    vector<vector<int>> merge(vector<vector<int>>& intervals) {
        sort(intervals.begin(), intervals.end(), 
            [](const vector<int>& a, const vector<int>& b){
                if (a[0] != b[0]) {
                    return a[0] < b[0];
                } 
                return a[1] < b[1];
        });

        vector<vector<int>> merged;

        merged.push_back(intervals[0]);
        for (int i = 1; i < intervals.size(); ++i) {
            auto& last = merged.back();
            if (last[1] < intervals[i][0]) {
                merged.push_back(intervals[i]); // not merge
            } else {
                last[1] = max(last[1], intervals[i][1]);
            }
        }
        return merged;
    }
};

File: 0056-merge-intervals/0056-merge-intervals.py
class Solution:
    def merge(self, intervals: List[List[int]]) -> List[List[int]]:
        intervals.sort(key=lambda x: x[0])

        merge_list = []

        for interval in intervals:
            if not merge_list:
                merge_list.append(interval)
            
            else:
                if interval[0] <= merge_list[-1][1]:
                    merge_list[-1][1] = max(merge_list[-1][1], interval[1])
                else:
                    merge_list.append(interval)

        return merge_list
        